src/test-utils/smoke-test-utils.ts

```diff
@@ -101,5 +153,50 @@ export const isEachEventIngested = async (
         if (!ingestedEvents.has(eventId)) {
             throw new Error(`Event ${eventId} not ingested.`);
         }
+
+        // validate custom attributes
+        if (metadataAttributes) {
```

Comment:
issue: This method is difficult to read due to its length and complexity. I suggest implementing the steps to retrieve the events, and then verify the events, as separate functions. See comment in the `verifyIngestionWithRetry` function.